sklearn/mixture/_gaussian_mixture_ic.py

```diff
@@ -0,0 +1,769 @@
+                )
+            if isinstance(self.affinity, list):
+                self.affinity.remove("cosine")
+            if self.affinity == "all":
+                self.affinity = ["euclidean", "manhattan", "none"]
+            warnings.warn(
+                "X contains a zero vector, will not run cosine affinity."
+            )
+
+        label_init = self.label_init
+        if label_init is not None:
+            if label_init.size != X.shape[0]:
+                msg = "n_samples must be the same as the length of label_init"
+                raise ValueError(msg)
+
+        param_grid = dict(
+            n_components=range(self.min_components, self.max_components + 1),
+            random_state=[self.random_state],
+        )
+        param_grid["affinity"] = (
+            self.affinity
+            if self.affinity != "all"
+            else ["euclidean", "manhattan", "cosine", "none"]
+        )
+        param_grid["linkage"] = (
+            self.linkage
+            if self.linkage != "all"
+            else ["ward", "complete", "average", "single"]
+        )
+        param_grid["covariance_type"] = (
+            self.covariance_type
+            if self.covariance_type != "all"
+            else ["spherical", "diag", "tied", "full"]
+        )
+
+        param_grid = list(ParameterGrid(param_grid))
+        param_grid_ag, param_grid = _process_paramgrid(param_grid)
+
+        n = X.shape[0]
+        if self.max_agglom_size is None or n <= self.max_agglom_size:
+            X_subset = X
+        else:  # if dataset is huge, agglomerate a subset
+            subset_idxs = np.random.choice(
+                np.arange(0, n), self.max_agglom_size
+            )
+            X_subset = X[subset_idxs, :]
+
+        ag_labels = []
+        if self.label_init is None:
+            for p_ag in param_grid_ag:
+                if p_ag["affinity"] != "none":
+                    agg = AgglomerativeClustering(
+                        n_clusters=self.min_components, **p_ag
+                    )
+                    agg.fit(X_subset)
+                    hierarchical_labels = _hierarchical_labels(
+                        agg.children_, self.min_components, self.max_components
+                    )
+                    ag_labels.append(hierarchical_labels)
+
+        def _fit_for_data(p):
+            ag_params = p[0]
+            gm_params = p[1]
+            n_clusters = gm_params["n_components"]
+            if (ag_params["affinity"] != "none") and (self.label_init is None):
+                index = param_grid_ag.index(ag_params)
+                agg_clustering = ag_labels[index][
+                    :, n_clusters - self.min_components
+                ]
+            else:
+                agg_clustering = []
+            return self._fit_cluster(X, X_subset, y, p, agg_clustering)
+
+        results = Parallel(n_jobs=self.n_jobs, verbose=self.verbose)(
+            delayed(_fit_for_data)(p) for p in param_grid
```

Comment:
You should be able to unwrap parameters here:
```suggestion
            delayed(_fit_for_data)(ag_params, gm_params) for (ag_params, gm_params) in param_grid
```

This can be propagated in the subsequent private methods signatures too.

Reply:
This suggestion also relate to others parts, such as:
 - https://github.com/scikit-learn/scikit-learn/pull/19562/files#diff-6c6a458c6e3e062e9487cb1d917ff2841086fdf47abd9c3588095af0decc513bR524-R526
  - https://github.com/scikit-learn/scikit-learn/pull/19562/files#diff-6c6a458c6e3e062e9487cb1d917ff2841086fdf47abd9c3588095af0decc513bR361-R369
   - https://github.com/scikit-learn/scikit-learn/pull/19562/files#diff-6c6a458c6e3e062e9487cb1d917ff2841086fdf47abd9c3588095af0decc513bR375-R382